domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -735,4 +743,164 @@ class ExplorationProgressController @Inject constructor(
       lastPlayedTimestamp
     )
   }
+
+  /**
+   * Updates the hint state in [ExplorationProgrses] with the one created from the saved checkpoint
+   * or the default instance of hint state if exploration is not being resumed.
+   */
+  private fun loadHintState(progress: ExplorationProgress) {
+    progress.hintState =
+      if (progress.currentExplorationCheckpoint == ExplorationCheckpoint.getDefaultInstance()) {
+        HintState.getDefaultInstance()
+      } else {
+        HintState.newBuilder().apply {
+          helpIndex = progress.currentExplorationCheckpoint.helpIndex
+          trackedAnswerCount = progress.currentExplorationCheckpoint.pendingUserAnswersCount
+          hintSequenceNumber = 0
+          delayToShowNextHintAndSolution = -1
+        }.build()
+      }
+  }
+
+  /**
+   * Initializes the variables of [StateDeck]. If the [ExplorationCheckpoint] is of type default
+   * instance, the values of [StateDeck] are reset. Otherwise, the variables of [StateDeck] are
+   * re-initialized with the values created from the saved [ExplorationCheckpoint].
+   *
+   * This function expects explorationProgress.hintState to be initialized with the correct values,
+   * so it should only be called after the function [loadHintState] has executed.
+   */
+  private fun loadStateDeck(progress: ExplorationProgress, exploration: Exploration) {
+    if (progress.currentExplorationCheckpoint == ExplorationCheckpoint.getDefaultInstance()) {
+      progress.stateDeck.resetDeck(progress.stateGraph.getState(exploration.initStateName))
+    } else {
+      progress.stateDeck.resumeDeck(
+        createPendingTopStateFromCheckpoint(progress),
+        getPreviousStatesFromCheckpoint(progress),
+        progress.currentExplorationCheckpoint.pendingUserAnswersList,
+        progress.currentExplorationCheckpoint.stateIndex
+      )
+    }
+  }
```

Comment:
This feels a bit like something that should be in the progress class itself since it's all about properly initializing the progress state. I suggest moving it over to keep this controller a bit more focused on its direct public API implementations.

Reply:
moved to exploration progress